### ISSUE TITLE
[Backend] Custom taxonomy in Posts Manage page doesn't translate

### ISSUE THREAD

**Issue by gmaggio** [not-only-code/qtranslate-slug#39]
I had a **custom taxonomy** that I had associated with my posts and I had created a **custom column** in the **posts manage page** to display the terms of this taxonomy in each posts.

However, **qTranslate Slug** doesn't seem to translate terms made from custom taxonomies. Built-in taxonomies (categories and tags) are fine in that they are translated.

**Comment by not-only-code** [comment]
Thanks for report.

**Comment by gmaggio** [comment]
Just to demonstrate, this was the code I used to display the **terms** from my **custom taxonomy** in the **post manage** page:

```
<?php
function my_custom_column($columns) {
  $columns['my_taxonomy'] = 'My Taxonomy';
  return $columns;
}
add_action('manage_posts_columns', 'my_custom_column', 10, 2);


function my_custom_column_output($column_name) {
  global $post;
  switch ($column_name) {
    case 'my_taxonomy':
      $terms = get_the_terms($post->ID, $column_name);
      if ( !empty( $terms ) ) {
        $terms_html = array();
        foreach ($terms as $term)
          array_push($terms_html, '<a href="edit.php?my_taxonomy='.$term->slug.'">'.$term->name.'</a>');        
        echo implode($terms_html, ", ");        
      };        
      break;
  }
}
add_action('manage_posts_custom_column', 'my_custom_column_output', 10, 2);
?>
```

The `$term->name` is **not** translated.

**Comment by gmaggio** [comment]
Okay, I managed to come up with a **workaround** and was able to **fix other issues as well**.

First, let me list down the number of issues I encountered in terms of backend taxonomy translations. **Most of the list below are not yet translated by qTranslate Slug** (**\* indicates the features already supported):

**1. Post Edit page:**
1. Category list 
2. Category parent dropdown menu
3. List of tags already added to the post
4. 'Most Used' tags list
5. Tags suggestions

**2. Post Manage page:**
1. Taxonomies ***
2. Custom taxonomies
3. Category filter dropdown menu

**3. Taxonomy Manage page:**
1. Popular tags
2. Category parent dropdown menu
3. Taxonomy names ***

![Taxonomy Translation Suggestions](https://f.cloud.github.com/assets/127763/377966/0bf6f226-a506-11e2-954a-8012499b4435.jpg)

Below is my workaround to fix almost all of the above issues. It may be a little rough, but it does the job. However, for the **tags in the Post Edit page**, despite successfully having them translated, when saved, if they're **not the default language**, Wordpress will create new ones. For this reason I had **disabled** the fix to this tags issue temporarily **(Issue 1.3 & 1.4)**.

Here's what I did:

## Remove some lines of code from the plugin

In **qtranslate-slug.php**, I removed line **438** and thus lines **1989-2020** as I don't think they would be necessary after my fix.

##### Line 438:

```
add_filter( 'term_name', array($this, 'term_name'), 0, 2 );
```

##### Lines 1989-2020:

From

```
/**
 * Bug fix for taxonomy names in wp_terms_list_table
 * 
```

to

```
  return $name;
}
```

## Then add these lines at the end:

```
if (is_admin()) {

  /**
   * Fix for:
   * - Taxonomy & custom taxonomy names in Post Manage page
   * - List of tags already added to the post in Post 
   *   Edit page (but have issues when saving)
   *   -> temporarily disabled
   */  
  function mod_wp_get_object_terms_qtranslate($terms, $obj_id, $taxonomy, $args) {

    global $pagenow;

    // Although in post edit page the tags are translated,
    // but when saving/updating the post Wordpess considers
    // the translated tags as new tags. Due to this
    // issue I limit this 'hack' to the post manage
    // page only.
    if ( $pagenow == 'edit.php' ) {

      // $taxonomy output seems to be wrapped
      // in single quotes, thus remove them to
      // make the output valid
      $tax = str_replace("'", "", $taxonomy);

      if ( !is_taxonomy_hierarchical($tax) ) {

        $meta    = get_option('qtranslate_term_name');
        $lang    = qtrans_getLanguage();

        if ( !empty( $terms ) ) {
          foreach ($terms as $term) {
            $term->original_name = $term->name;
            $term->name = $meta[$term->name][$lang];
          };
        };

      }

    }
    return $terms;
  }
  add_filter( 'wp_get_object_terms', 'mod_wp_get_object_terms_qtranslate', 0, 4 );


  /**
   * Fix for:
   * - Taxonomy names in Taxonomy Manage page
   * - 'Popular Tags' in Taxonomy (Tags) Manage page
   * - Category filter dropdown menu in Post Manage page
   * - Category list in Post Edit page
   * - 'Most Used' tags list in Post Edit page
   *   (but have issues when saving)
   *   -> temporarily disabled
   */  
  function mod_get_terms_qtranslate($terms, $taxonomy) {

    global $pagenow;

    // Although in post edit page the tags in 'most
    // used' list are translated, but when saving the
    // post Wordpess considers the translated tags as
    // new tags. Due to this issue I skip this 'hack'
    // for tags in post edit page.
    if ( $pagenow != 'admin-ajax.php' ) {

      $meta    = get_option('qtranslate_term_name');
      $lang    = qtrans_getLanguage();

      if ( !empty( $terms ) ) {
        foreach ($terms as $term) {
          if ($meta[$term->name][$lang]) {
            $term->name = $meta[$term->name][$lang];
          }
        };
      };

    }

    return $terms;
  }
  add_filter( 'get_terms', 'mod_get_terms_qtranslate', 0, 3 );

}
```

I have yet to find a way to translate the **tags suggestions**, though **(Issue 1.5)**.

## Checklist

These are the list of what my code fixes:

**1. Post Edit page:**
- [x] Category list 
- [x] Category parent dropdown menu
- [ ] List of tags already added to the post **[works but have issues -> DISABLED]**
- [ ] 'Most Used' tags list **[works but have issues -> DISABLED]**
- [ ] Tags suggestions

**2. Post Manage page:**
- [x] Taxonomies
- [x] Custom taxonomies
- [x] Category filter dropdown menu

**3. Taxonomy Manage page:**
- [x] Popular tags
- [x] Category parent dropdown menu
- [x] Taxonomy names

**Comment by not-only-code** [comment]
Hi, i'm busy now, but when ill have some time, ill check this. Thanks for your report

**Comment by LC43** [comment]
i think it has been fixed in 6311c838636bc1bb2348ccacd932448f55a9f8d6
